README.md
# Turn and Run

**This software should be considered alpha quality. It works fine for
me, but has not been used by other people. Please report any issues.**

## A radio station selector dial

The turnandrun program was written to provide a manual interface
to an internet radio player running on a Raspberry Pi. It manages
a single dial that is turned to marked positions to select stations to
play, or to stop the audio.

These instructions describe a 10k potentiometer connected to an ADS1115
or ADS1015 ADC, and installation on Moode (or Raspberry Pi OS). However,
it should be possible to adapt them to another input device, or another
ADC, or another operating system.

A simple configuration file specifies commands to be run at certain
positions of the dial. These commands can be anything.

## Install the hardware

### Connect the hardware
The software should work with ADS1115 and ADS1015 modules. I have an
ADS1115 module and 10k potentiometer, wired like the following diagram
(in fact, I have a cheap fake ADS1115 module, that is really an ADS1015,
and counts in steps of 16, but is fine for use with turnandrun)

![ADS1115 wiring diagram](wiring_ads1x15.png)

Note: the wiring diagram connects ADDR to GND, which onfigures the board to
use default I2C address 0x48.

### Configure the system

Enable the kernel driver for the first channel of the ADS1X15 board.
Edit the boot configuration file
```
sudo nano /boot/config.txt
```
For an ADS1115, add the line
```
dtoverlay=ads1115,cha_enable,cha_gain=1
```
For an ADS1015, add the line
```
dtoverlay=ads1015,cha_enable,cha_gain=1
```

Ensure the I2C module is loaded at boot. Edit the modules file
```
sudo nano /etc/modules
```
Add the following line, if it is not already included
```
i2c-dev
```

Reboot to activate these changes
```
sudo reboot
```


## Install turnandrun program

The only installation method currenly available is to build and install
from source.

### Build and install from Source

Install the packages needed
```
sudo apt update
sudo apt install build-essential autoconf libtool libiio-dev i2c-tools
```

Download, build and install turnandrun
```
git clone https://github.com/antiprism/turnandrun
cd turnandrun
./bootstrap
CPPFLAGS="-W -Wall -Wno-psabi" ./configure
make
sudo make install-strip
```

## Configure the program

The program is configured using a simple text file. The default
configuration file is `/etc/turnandrun.conf` (or specify a different
file, e.g. for testing, with `turnandrun -c path_to_conf_file`).

Edit the defaul configuration file and add your settings
```
sudo nano /etc/turnandrun.conf
```

### Configure dial commands

Use marks around the dial as positions where you would like commands
to be executed. These positions correspond to a raw dial dial reading,
which can be associated with a command to run in a confguration file.

To see the raw readings, us the dry run option
```
turnandrun -d
```
Make a record of the raw values corresponding to marks where you
want to run commands.

Each command is added to the configuration file as a single line in
the form `raw_reading=comand_label,command`, where `raw_reading` is the
raw number, `command_label` is short label to identify the command
(cannot include a comma), and `command` is the text of
the command to run. En example command line might be
```
24000 = stop, mpc stop
```

### Configure other settings

These settings only need to be included in the configuration file to
change the default value. The settings are added one on each line in
the form `setting=value`.

**channel = letter** (default: a, valid: a, b, c, d):
This is the channel on the ADS1X115 board that the potentiometer is
conected to. Channel 'a' is the first channel, perhaps marked 'A0'
on the board, and enabled in config.txt with 'cha_enable,cha_gain=1'.
Channel 'b' is the second channel, perhaps marked 'A1'
on the board, and enabled in config.txt with 'chb_enable,chb_gain=1'.
Likewise for channel values 'c' and 'd'.

**turn_before_run = bool** (default: 1, valid: 0, 1):
Specifies if the command corresponding to the current dial position
should be run on startup - '0' run the command, '1' wait for the dial
to change position before running any command. Don't
change from the default unless you are happy for any of the commands
to be run when the machine is started up, for example, when it is plugged
it in, or when it starts up after a power cut in the night.

**command_delay = seconds** (default: 1, range: 0 - 10):
Number of seconds for the dial to be at a mark before running the
command. The delay stops commands from being executed when the dial
is turned through one mark to get to another.

**frequency = per_second** (default: 10 range: 1 - 100)
Number of times per second to read the dial position.

**dead_zone_width = percent** (default: 5, range: 0 - 50)
Half-way between two adjacent dial marks is a dead zone, which is not
associated with any mark. The setting is the percentage of the width
between the two marks that should be used as the dead zone. The dead
zone eans that marks are clearly differentiated, and the dial cannot
inintentionally wander between marks when it is close to the mid-point.

Example configuration lines
```
turn_before_run = 1
dead_zone = 5
command_delay = 1
channel = a  
frequency = 10
```

### Example configuration file

This is my radio player configuration. I use defaults for the settings
(but specify the channel for the example), and specify commands for
four dial positions. The first dial position stops the audio, and the
other three positions play BBC Radio 1, 2 and 4.
```
channel = a  
0 = stop, mpc -q stop
8000 = BBC Radio 1, mpc -q clear && mpc -q add http://stream.live.vc.bbcmedia.co.uk/bbc_radio_one && mpc -q play
16000 = BBC Radio 2, mpc -q clear && mpc -q add http://stream.live.vc.bbcmedia.co.uk/bbc_radio_two && mpc -q play
24000 = BBC Radio 4, mpc -q clear && mpc -q add http://stream.live.vc.bbcmedia.co.uk/bbc_radio_fourfm && mpc -q play

```

### Testing the configuration

Run the program with option `-d` for a dry run.
```
turnandrun -d
```
This will produce a report on the screen of: the configuration options,
the current dial value and mark it corresponds to, and each time a command
would be run.

### Installing the service

After installing the service the turnandrun program will run
when the machine starts. Install the service with
```
sudo turnandrun_service_install
```
(and if you later want to uninstall the servicew run
`sudo turnandrun_service_uninstall`)


## Program Help and Options

The following text is printed by running `turnandrun -h`

```
Usage: turnandrun [options]

A 10k potentiometer is connected to an ADS1115 or ADS1015 ADC and a
simple configuration file determines commands to run at certain
positions.

Options
  -h,--help this help message
  --version version information

  -c <file>  configuration file name (default: /etc/turnandrun.conf), may
             contain the following lines (with two or more command lines)

               dial_reading_number = command_label, command_to_run

               channel = letter             (default: a, valid: a, b, c, d)
               turn_before_run = bool       (default: 1, valid: 0, 1)
               command_delay = seconds      (default: 1, range: 0 - 10)
               frequency = per_second       (default: 10 range: 1 - 100)
               dead_zone = percent_band    (default: 5, range: 0 - 50)
  -d         dry run, no commands are run, print a report to the screen of
             the configuration options, the current dial value and mark it
             corresponds to, and each time a command would be run.

```



NEWS
Version 0.01
xx-05-21 Adrian Rossiter <adrian@antiprism.com>
         Changes
           First release.


src/dial.h
/*
   Copyright (c) 2021, Adrian Rossiter

   Antiprism - http://www.antiprism.com

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the "Software"),
   to deal in the Software without restriction, including without limitation
   the rights to use, copy, modify, merge, publish, distribute, sublicense,
   and/or sell copies of the Software, and to permit persons to whom the
   Software is furnished to do so, subject to the following conditions:

      The above copyright notice and this permission notice shall be included
      in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
  IN THE SOFTWARE.
*/

#ifndef DIAL_H
#define DIAL_H

#include "status_msg.h"

#include <limits>
#include <map>
#include <string>
#include <vector>

class DialBands {
public:
  static const long unset = std::numeric_limits<long>::min();
  void add_band(long k, long v) { bands[k] = v; }
  long get_mark(long val) const;
  const std::map<long, long> &get_bands() const { return bands; }

private:
  std::map<long, long> bands;
};

class DialSettings {
public:
  struct Command {
    std::string label;
    std::string command;
  };

  Command get_command(long dial_reading) const;
  Status set_command(int dial_reading, std::string cmd_label,
                     std::string cmd_command);
  Status set_setting(std::string setting, std::string value);

  char get_channel() const { return channel; }
  bool get_turn_before_run() const { return turn_before_run; }
  double get_command_delay() const { return command_delay; }
  double get_dead_zone() const { return dead_zone; }
  double get_frequency() const { return frequency; }
  int get_channel_num() const { return channel - 'a'; }
  std::map<long, Command> get_commands() const { return commands; }
  DialBands create_dial_bands(double dead_frac) const;
  std::string dial_bands_report(const DialBands &dial_bands) const;

  std::string settings_report() const;
  void set_dry_run(bool dry = true) { dry_run = dry; }
  bool get_dry_run() const { return dry_run; }

private:
  bool turn_before_run = true; // turn dial before first command is run
  char channel = 'a';          // AD1X15 channel (a, b, c, or d)
  double command_delay = 1;    // secs stopped before command is run
  double dead_zone = 0.05;     // dead fraction between bands
  double frequency = 10.0;     // polling frequency
  std::map<long, Command> commands;

  bool dry_run = false; // verbose output, do not run commands
};

#endif // DIAL_H


src/utils.h
/*
   Copyright (c) 2003-2018, Adrian Rossiter

   Antiprism - http://www.antiprism.com

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the "Software"),
   to deal in the Software without restriction, including without limitation
   the rights to use, copy, modify, merge, publish, distribute, sublicense,
   and/or sell copies of the Software, and to permit persons to whom the
   Software is furnished to do so, subject to the following conditions:

      The above copyright notice and this permission notice shall be included
      in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
  IN THE SOFTWARE.
*/

/*!\file utils.h
   \brief utility routines for text operations, I/O conversions, etc
*/

#ifndef UTILS_H
#define UTILS_H

#include "status_msg.h"

#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <vector>

/// Read a floating point number, or mathematical expression, from a string.
/** The string should only hold the floating point number or expression, but may
 *  have leading and trailing whitespace.
 * \param str the string holding the floating point number or expressions.
 * \param f used to return the floating point number.
 * \return status, evaluates to \c true if a valid floating point number
 *  was read, otherwise \c false.*/
Status read_double(const char *str, double *f);

/// Read an integer from a string.
/** The string should only hold the integer, but may
 *  have leading and trailing whitespace.
 * \param str the string holding the integer.
 * \param i used to return the integer.
 * \return status, evaluates to \c true if a valid integer
 *  was read, otherwise \c false.*/
Status read_int(const char *str, int *i);

/// Join strings into a single string, with parts separated by a delimiter
/**\param iterator to first string in sequence
 * \param iterator to end of string sequence
 * \param delim delimiter to include between strings
 * \return The joined and delimited strings */
template <typename InputIt>
std::string join(InputIt first, InputIt last, std::string delim = " ")
{
  std::string str;
  for (; first != last; ++first)
    str += *first + ((std::next(first) != last) ? delim : std::string());
  return str;
}

/// Convert a C formated message string to a C++ string
/** Converts the first MSG_SZ-1 characters of the C format string
 * \param fmt the formatted string
 * \param ... the values for the format
 * \return The converted string. */
std::string msg_str(const char *fmt, ...);

#endif // UTILS_H


src/dial.cpp
/*
   Copyright (c) 2021, Adrian Rossiter

   Antiprism - http://www.antiprism.com

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the "Software"),
   to deal in the Software without restriction, including without limitation
   the rights to use, copy, modify, merge, publish, distribute, sublicense,
   and/or sell copies of the Software, and to permit persons to whom the
   Software is furnished to do so, subject to the following conditions:

      The above copyright notice and this permission notice shall be included
      in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
  IN THE SOFTWARE.
*/

#include "dial.h"
#include "utils.h"

#include <cstring>

using std::string;

long DialBands::get_mark(long kval) const
{
  auto it_after = bands.upper_bound(kval);
  if (bands.size() == 0)
    return unset;
  else if (it_after == bands.end())   // kval is beyond the highest key
    return bands.rbegin()->second;    // largest mark;
  else if (it_after == bands.begin()) // kval is before lowest mark
    return bands.begin()->second;
  else
    return (--it_after)->second; // use start of including band
}

Status DialSettings::set_command(int dial_reading, std::string cmd_label,
                                 std::string cmd_command)
{
  if (dial_reading < 0)
    return Status::error("dial reading cannot be negative");

  if (cmd_label.empty())
    return Status::error("no command label given");

  if (cmd_command.empty())
    return Status::error("no command given");

  commands[dial_reading] = {cmd_label, cmd_command};

  return Status::ok();
}

DialSettings::Command DialSettings::get_command(long dial_reading) const
{
  auto it = commands.find(dial_reading);
  return (it == commands.end()) ? Command() : it->second;
}

Status DialSettings::set_setting(std::string setting, std::string value)
{
  if (setting.empty())
    return Status::error("no setting given");

  if (value.empty())
    return Status::error("no value given");

  string msg_prefix = "setting '" + setting + "': ";
  if (setting == "dead_zone" || setting == "command_delay" ||
      setting == "frequency") {
    string msg_prefix2 = msg_prefix + "value '" + value + "': ";
    double num;
    Status stat = read_double(value.c_str(), &num);
    if (!stat)
      return Status::error(msg_prefix2 + "not a number: " + stat.msg());

    int lim_low = (setting == "frequency") ? 1 : 0;
    int lim_high = (setting == "command_delay") ? 10 : 50;
    if (num < lim_low || num > lim_high)
      return Status::error(msg_prefix2 + "must be in range " +
                           std::to_string(lim_low) + " to " +
                           std::to_string(lim_high));

    if (setting == "dead_zone")
      dead_zone = num / 100;
    else if (setting == "command_delay")
      command_delay = num;
    else // setting == frquency
      frequency = num;
  }
  else if (setting == "channel") {
    if (value.size() != 1 || !strchr("abcd", value[0]))
      return Status::error(msg_prefix +
                           "must be one letter, from a, b, c or d");
    channel = value[0];
  }
  else if (setting == "turn_before_run") {
    if (value.size() != 1 || !strchr("01", value[0]))
      return Status::error(msg_prefix + "must be one digit, 0 or 1");
    turn_before_run = (value[0] == '1');
  }
  else
    return Status::error(msg_prefix + "unknown setting");

  return Status::ok();
}

DialBands DialSettings::create_dial_bands(double dead_frac) const
{
  DialBands dial_bands;

  if (get_commands().size() >= 2) {
    long cur_mark;
    long prev_mark = DialBands::unset;
    for (auto kp : get_commands()) {
      cur_mark = kp.first;
      if (prev_mark == DialBands::unset) { // cur_mark is first mark
        dial_bands.add_band(cur_mark, cur_mark);
      }
      else {
        long mid_point = (prev_mark + cur_mark) / 2;
        if (dead_frac > 0) {
          long dead_zone = (cur_mark - prev_mark) * dead_frac;
          dial_bands.add_band(mid_point - dead_zone / 2, DialBands::unset);
          dial_bands.add_band(mid_point + dead_zone / 2, cur_mark);
        }
        else
          dial_bands.add_band(mid_point, cur_mark);
      }
      prev_mark = cur_mark;
    }

    dial_bands.add_band(prev_mark, prev_mark);
  }

  return dial_bands;
};

std::string DialSettings::dial_bands_report(const DialBands &dial_bands) const
{
  string str;
  const auto &bands = dial_bands.get_bands();
  const auto unset = DialBands::unset;

  if (bands.size() < 2)
    str = "no command bands (need at least two commands)\n";
  else {
    auto band = bands.begin();
    auto next = bands.end();
    auto next_again = bands.begin();
    while (next_again != bands.end()) {
      string start_str =
          msg_str("%2s%7ld", (next == bands.end()) ? "<" : "", band->first);

      next = std::next(band);
      next_again = (next == bands.end()) ? bands.end() : std::next(next);

      string end_str = (next_again == bands.end())
                           ? msg_str("%7ld%2s", next->first, ">")
                           : msg_str("%7ld%2s", next->first - 1, "");

      auto cmd = get_command(band->second);
      string cmd_str = (band->second == unset) ? "-------------" : cmd.label;
      str += "(" + start_str + " , " + end_str + ") : " + cmd_str + "\n";
      band = next;
    }
  }
  return str;
}

std::string DialSettings::settings_report() const
{
  string str;
  str += msg_str("turn_before_run = %d\n", turn_before_run);
  str += msg_str("channel = %c\n", channel);
  str += msg_str("command_delay = %g\n", command_delay);
  str += msg_str("dead_zone = %g\n", dead_zone);
  str += msg_str("frequency = %g\n", frequency);
  str += "\n";
  for (auto kp : commands)
    str += msg_str("%7ld = %s, %s\n", kp.first, kp.second.label.c_str(),
                   kp.second.command.c_str());

  return str;
}


src/utils.cpp
/*
   Copyright (c) 2003-2018, Adrian Rossiter

   Antiprism - http://www.antiprism.com

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the "Software"),
   to deal in the Software without restriction, including without limitation
   the rights to use, copy, modify, merge, publish, distribute, sublicense,
   and/or sell copies of the Software, and to permit persons to whom the
   Software is furnished to do so, subject to the following conditions:

      The above copyright notice and this permission notice shall be included
      in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
  IN THE SOFTWARE.
*/

/* \file utils.cpp
   \brief utility routines for maths operations, text operations,
   I/O conversions, etc
*/
#include "utils.h"

#include <ctype.h>
#include <limits.h>
#include <math.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <sys/stat.h>
#include <sys/types.h>
#include <vector>

using std::string;
using std::vector;

Status read_double(const char *str, double *f)
{
  bool to_sqrt;
  char buff;
  if (sscanf(str, " sqrt%lf %c", f, &buff) == 1)
    to_sqrt = true;
  else if (sscanf(str, " %lf %c", f, &buff) == 1)
    to_sqrt = false;
  else
    return Status::error("not a number");

  if (isinf(*f))
    return Status::error("number too large\n");

  if (isnan(*f))
    return Status::error("not a number\n");

  if (to_sqrt)
    *f = sqrt(*f);

  return Status::ok();
}

Status read_int(const char *str, int *i)
{
  char buff;
  if (sscanf(str, " %d %c", i, &buff) != 1)
    return Status::error("not an integer");

  if (*i == INT_MAX)
    return Status::error("integer too large\n");

  return Status::ok();
}

// https://stackoverflow.com/questions/2342162/stdstring-formatting-
// like-sprintf/49812018#49812018
string msg_str(const char *fmt, ...)
{
  // initialize use of the variable argument array
  va_list ap;
  va_start(ap, fmt);

  // reliably acquire the size
  // from a copy of the variable argument array
  // and a functionally reliable call to mock the formatting
  va_list ap_copy;
  va_copy(ap_copy, ap);
  const int len = vsnprintf(NULL, 0, fmt, ap_copy);
  va_end(ap_copy);

  // return a formatted string without risking memory mismanagement
  // and without assuming any compiler or platform specific behavior
  vector<char> vstr(len + 1);
  vsnprintf(vstr.data(), vstr.size(), fmt, ap);
  va_end(ap);
  return string(vstr.data(), len);
}


src/main.cpp
/*
   Copyright (c) 2021, Adrian Rossiter

   Antiprism - http://www.antiprism.com

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the "Software"),
   to deal in the Software without restriction, including without limitation
   the rights to use, copy, modify, merge, publish, distribute, sublicense,
   and/or sell copies of the Software, and to permit persons to whom the
   Software is furnished to do so, subject to the following conditions:

      The above copyright notice and this permission notice shall be included
      in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
  IN THE SOFTWARE.
*/

#include "dial.h"
#include "programopts.h"
#include "timer.h"
#include "utils.h"

#include <iio.h>

#include <unistd.h>

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <limits>
#include <map>
#include <string>
#include <vector>

using std::string;

class DialOpts : public ProgramOpts {
public:
  DialSettings dial_settings;
  std::string config_file_name = "/etc/turnandrun.conf"; // config file name
  bool dry_run = false;
  bool just_raw_values = false;

  DialOpts() : ProgramOpts("turnandrun", "0.01") {}
  void process_command_line(int argc, char **argv);
  void usage();
};

void DialOpts::usage()
{
  fprintf(stdout, R"(
Usage: %s [options]

A 10k potentiometer is connected to an ADS1115 or ADS1015 ADC and a
simple configuration file determines commands to run at certain
positions.

Options
%s
  -c <file>  configuration file name (default: /etc/turnandrun.conf), may
             contain the following lines (with two or more command lines)

               dial_reading_number = command_label, command_to_run

               channel = letter             (default: a, valid: a, b, c, d)
               turn_before_run = bool       (default: 1, valid: 0, 1)
               command_delay = seconds      (default: 1, range: 0 - 10)
               frequency = per_second       (default: 10 range: 1 - 100)
               dead_zone = percent_band    (default: 5, range: 0 - 50)
  -d         dry run, no commands are run, print a report to the screen of
             the configuration options, the current dial value and mark it
             corresponds to, and each time a command would be run.
)",
          get_program_name().c_str(), help_ver_text);
}

void DialOpts::process_command_line(int argc, char **argv)
{
  opterr = 0;
  int c;

  handle_long_opts(argc, argv);

  while ((c = getopt(argc, argv, ":hc:d")) != -1) {
    if (common_opts(c, optopt))
      continue;

    switch (c) {
    case 'c':
      config_file_name = optarg;
      break;

    case 'd':
      dry_run = true;
      break;

    default:
      error("unknown command line error");
    }
  }

  if (argc - optind > 0)
    error(msg_str("invalid option or parameter: '%s'", argv[optind]));
}

int read_line(FILE *file, char **line)
{

  int linesize = 128;
  *line = (char *)malloc(linesize);
  if (!*line)
    return -1;

  int offset = 0;
  while (true) {
    if (!fgets(*line + offset, linesize - offset, file)) {
      if (offset != 0)
        return 0;
      else {
        *(*line + offset) = '\0'; // terminate the line
        return (ferror(file)) ? -1 : 1;
      }
    }
    int len = offset + strlen(*line + offset);
    if ((*line)[len - 1] == '\n') {
      (*line)[len - 1] = 0;
      return 0;
    }
    offset = len;

    char *newline = (char *)realloc(*line, linesize * 2);
    if (!newline)
      return -1;
    *line = newline;
    linesize *= 2;
  }
}

// Trim whitespace - https://stackoverflow.com/a/17976541
std::string trim(const std::string &s)
{
  auto wsfront = std::find_if_not(s.begin(), s.end(),
                                  [](int c) { return std::isspace(c); });
  auto wsback = std::find_if_not(s.rbegin(), s.rend(),
                                 [](int c) { return std::isspace(c); })
                    .base();
  return (wsback <= wsfront ? std::string() : std::string(wsfront, wsback));
}

Status read_config_file(const string &file_name, DialSettings &dial_settings)
{
  dial_settings = DialSettings();

  FILE *file = fopen(file_name.c_str(), "r");
  if (file == NULL)
    return Status::error("could not open file");

  // config file has two kinds of lines:
  //    setting = value
  //    dial_reading = command_id , command

  char *line;
  int line_no = 0;
  int ret;
  while ((ret = read_line(file, &line)) == 0) {
    line_no++;
    auto line_str = trim(line); // to ignore whitespace-only lines
    free(line);                 // memory freed before any return

    if (!line_str.empty()) {
      string msg_prefix_line = "line " + std::to_string(line_no) + ": ";

      // split line on first '='
      auto pos_equal = line_str.find('=');
      if (pos_equal == string::npos)
        return Status::error(msg_prefix_line + "did not include '='");

      auto setting = trim(line_str.substr(0, pos_equal));

      // check if setting is setting string or dial reading number
      int dial_reading;
      if (read_int(setting.c_str(), &dial_reading)) {
        // line: dial_reading = command_id , command
        string msg_prefix_cmd = msg_prefix_line + "dial command: ";

        // split value on first ','
        auto pos_comma = line_str.find(',', pos_equal + 1);
        if (pos_comma == string::npos)
          return Status::error(msg_prefix_cmd + "did not include a ','");

        // label before first ','
        string cmd_label =
            trim(line_str.substr(pos_equal + 1, pos_comma - pos_equal - 1));

        // command after first ','
        string cmd_command = trim(line_str.substr(pos_comma + 1));

        Status stat =
            dial_settings.set_command(dial_reading, cmd_label, cmd_command);
        if (!stat)
          return Status::error(msg_prefix_line + "dial command: " + stat.msg());
      }
      else {
        // line: setting = value
        auto value = trim(line_str.substr(pos_equal + 1));
        Status stat = dial_settings.set_setting(setting, value);
        if (!stat)
          return Status::error(msg_prefix_line + "setting: " + stat.msg());
      }
    }
  }
  fclose(file);

  if (dial_settings.get_commands().size() < 2)
    return Status::error("included less than two commands");

  if (ret < 0)
    return Status::error("memory allocation error while reading file");

  return Status::ok();
}

Status loop(iio_device *device, const DialSettings &dial_settings)
{
  string attr_v_raw =
      "in_voltage" + std::to_string(dial_settings.get_channel_num()) + "_raw";

  auto dial_bands =
      dial_settings.create_dial_bands(dial_settings.get_dead_zone());

  if (dial_settings.get_dry_run()) {
    printf("\n== Configuration Settings ==\n");
    printf("%s\n", dial_settings.settings_report().c_str());
    printf("\n== Band Settings ==\n");
    printf("%s\n", dial_settings.dial_bands_report(dial_bands).c_str());
  }

  // dial position mark that was last stopped on
  long mark_stop = DialBands::unset;

  // dial postion mark for last raw value
  long mark_last = DialBands::unset;

  // check whether to execute current command on start, or wait for dial change
  // (by setting a long intial delay on the same band before running command)
  double initial_delay = (dial_settings.get_turn_before_run())
                             ? 10000000                           // long time
                             : dial_settings.get_command_delay(); // usual time
  Timer timer(initial_delay);

  if (dial_settings.get_dry_run())
    printf("\n== Processing Loop ==\n");

  bool first_loop = true;
  while (true) {
    long long raw;
    if (iio_device_attr_read_longlong(device, attr_v_raw.c_str(), &raw) != 0) {
      return Status::error("could not read values from ADS1X15 device from " +
                           attr_v_raw);
    }

    auto mark_now = dial_bands.get_mark(raw); // mark for current raw value
    if (first_loop) {
      mark_last = mark_now; // initially no change
      first_loop = false;
    }

    // If current mark has changed then restart the timer
    if (mark_now != mark_last)
      timer.set_timer(dial_settings.get_command_delay());

    // check if the dial...
    //    has been in the current band for the delay time, AND
    //    is no longer in the last stop band, AND
    //    is not in a dead zone
    if (timer.finished() && mark_now != mark_stop &&
        mark_now != DialBands::unset) {
      // dial has stopped in a new band
      mark_stop = mark_now;
      auto cmd = dial_settings.get_command(mark_stop);
      if (dial_settings.get_dry_run())
        printf("COMMAND (mark: %-10ld) %s: %s\n", mark_stop, cmd.label.c_str(),
               cmd.command.c_str());
      else
        system(cmd.command.c_str());
    }

    if (dial_settings.get_dry_run()) {
      string str = msg_str("raw: %-7lld, mark: ", raw);
      if (mark_now == DialBands::unset)
        str += "dead zone";
      else {
        auto label = dial_settings.get_command(mark_now).label;
        str += msg_str("%-7ld, label: %s ", mark_now,
                       label.substr(0, std::min(40u, label.size())).c_str());
      }
      printf("%-70s\r", str.c_str());
      fflush(stdout);
    }

    mark_last = mark_now;

    usleep(1000000 / dial_settings.get_frequency());
  }

  return Status::ok();
}

int main(int argc, char **argv)
{
  DialOpts opts;
  opts.process_command_line(argc, argv);

  DialSettings dial_settings;
  Status stat = read_config_file(opts.config_file_name, dial_settings);
  if (opts.dry_run) {
    dial_settings.set_dry_run();
    // print result but don't exit on error
    printf("\n== Read Configuration File ==\n");
    printf("File name: '%s': ", opts.config_file_name.c_str());
    if (stat.is_ok())
      printf("ok");
    else if (stat.is_warning())
      printf("warning: %s", stat.c_msg());
    else
      printf("error: %s", stat.c_msg());
    printf("\n");
  }
  else
    opts.print_status_or_exit(stat,
                              "config file '" + opts.config_file_name + "'");

  auto iio_context = iio_create_local_context();
  auto device = iio_context_find_device(iio_context, "ads1015");
  if (device == nullptr)
    opts.error("could not open ADS1X15 device");

  opts.print_status_or_exit(loop(device, dial_settings));

  return 0;
}
